In [20]:
from scipy.io import loadmat
import numpy as np
import pandas as pd


In [21]:


# Load the .mat file
mat_data = loadmat('influenza_outbreak_dataset.mat')

# Print the keys in the dictionary
print(mat_data.keys())



dict_keys(['__header__', '__version__', '__globals__', 'flu_X_tr', 'flu_Y_tr', 'flu_X_te', 'flu_Y_te', 'flu_locs', 'flu_keywords'])


In [22]:


# Given array of arrays of keywords
flu_keywords = np.array(mat_data['flu_keywords'][0])

# Initialize an empty list to store concatenated keywords
all_keywords = []

# Iterate over each array of keywords and concatenate them
for keyword_array in flu_keywords:
    all_keywords.extend(keyword_array)

# Convert the list to a numpy array
all_keywords_array = np.array(all_keywords)

print(all_keywords_array)
print(len(all_keywords_array))



['flu' 'swine' 'stomach' 'symptoms' 'virus' 'bug' 'strep' 'season'
 'influenza' 'fever' 'thera' 'poisoning' 'pneumonia' 'bird' 'infection'
 'epidemic' 'week' 'bronchitis' 'sucks' 'immune' 'soon' 'colds' 'caught'
 'medicine' 'soup' 'weekend' 'dying' 'recovering' 'weeks' 'meds' 'sick'
 'outbreak' 'cough' 'sickness' 'strain' 'jab' 'kicking' 'catch' 'catching'
 'bed' 'contagious' 'days' 'aches' 'yesterday' 'battling' 'cure' 'coming'
 'doctor' 'rest' 'worst' 'tomorrow' 'case' 'illness' 'hours' 'due' 'worse'
 'system' 'ini' 'weight' 'sinus' 'body' 'germs' 'morning' 'hangover'
 'death' 'awful' 'hospital' 'fighting' '12s' 'away' 'feeling' 'sakit'
 'horrible' 'miserable' 'since' 'ebola' 'lagi' 'die' 'chills' 'finals'
 'vaccination' 'pounds' 'weather' 'nasty' 'played' 'hour' 'healthy'
 'sicker' 'winter' 'finally' 'plague' 'spreading' 'food' 'avian' 'tablets'
 'joke' 'pigs' 'severe' 'ridden' 'missed' 'terrible' 'puking' 'disease'
 'ago' 'mono' 'coughing' 'tea' 'vaccine' 'office' 'ill' 'deaths' 's

In [23]:
import numpy as np

# Given array of arrays of keywords
flu_locs = np.array(mat_data['flu_locs'][0])

# Initialize an empty list to store concatenated keywords
all_locs = []

# Iterate over each array of keywords and concatenate them
for loc_array in flu_locs:
    all_locs.extend(loc_array)

# Convert the list to a numpy array
all_locs_array = np.array(all_locs)

print(all_locs_array)
print(len(all_locs_array))



['wyoming' 'colorado' 'nebraska' 'washington' 'rhode island' 'tennessee'
 'wisconsin' 'nevada' 'maine' 'north dakota' 'mississippi' 'new jersey'
 'oklahoma' 'delaware' 'minnesota' 'north carolina' 'illinois' 'new york'
 'arkansas' 'indiana' 'maryland' 'louisiana' 'idaho' 'arizona' 'iowa'
 'west virginia' 'michigan' 'kansas' 'utah' 'virginia' 'oregon'
 'connecticut' 'california' 'massachusetts' 'alaska' 'vermont' 'georgia'
 'pennsylvania' 'florida' 'hawaii' 'district of columbia' 'new hampshire'
 'texas' 'missouri' 'south carolina' 'ohio' 'alabama' 'new mexico']
48


In [24]:
# Initialize an empty DataFrame
X_train = pd.DataFrame()

# Iterate over each element in data['flu_X_tr'][0]
for index, element in enumerate(mat_data['flu_X_tr'][0]):
    dense_matrix = element.toarray()
    df = pd.DataFrame(dense_matrix)
    df['target'] = mat_data['flu_Y_tr'][0][index].flatten()
    df['state'] = all_locs_array[index]
    df['week'] = df.index+1

    # Concatenate df with merged_df
    X_train = pd.concat([X_train, df])

# Resetting the index of the merged DataFrame
X_train.reset_index(drop=True, inplace=True)

# Display the DataFrame
print(X_train)


         0    1    2    3    4    5    6    7    8    9  ...  538  539  540  \
0      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0   
1      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
52555  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  0.0  0.0  0.0   
52556  1.0  1.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  ...  0.0  1.0  0.0   
52557  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
52558  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  ...  0.0  0.0  0.0   
52559  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

       541  542  543  544  target       state  week

## Decsion tree classifier:

In [30]:
from sklearn.tree import DecisionTreeClassifier
import pandas as pd

# One-hot encode the 'state' column
X_train_encoded = pd.get_dummies(X_train, columns=['state'])

# Initialize the Decision Tree classifier
dt_classifier = DecisionTreeClassifier()

# Extract features (X) and target (y) from the DataFrame
X = X_train_encoded.drop(columns=['target'])  # Remove non-feature columns
y = X_train_encoded['target']

# Convert feature names to strings
X.columns = X.columns.astype(str)

# Train the Decision Tree classifier
dt_classifier.fit(X, y)


DecisionTreeClassifier()

In [40]:
# Make predictions on the training data
y_train_pred = dt_classifier.predict(X)

# Calculate training accuracy
train_accuracy = accuracy_score(y, y_train_pred)
print("Training Accuracy:", train_accuracy)


Training Accuracy: 1.0


In [34]:
# Initialize an empty DataFrame
X_test = pd.DataFrame()

# Iterate over each element in data['flu_X_tr'][0]
for index, element in enumerate(mat_data['flu_X_te'][0]):
    dense_matrix = element.toarray()
    df = pd.DataFrame(dense_matrix)
    df['target'] = mat_data['flu_Y_te'][0][index].flatten()
    df['state'] = all_locs_array[index]
    df['week'] = df.index+1

    # Concatenate df with merged_df
    X_test = pd.concat([X_test, df])

# Resetting the index of the DataFrame
X_test.reset_index(drop=True, inplace=True)

X_test.columns = X_test.columns.astype(str)
# Display the DataFrame
print(X_test)



         0    1    2    3    4    5    6    7    8    9  ...  538  539  540  \
0      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
23275  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
23276  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
23277  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
23278  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
23279  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

       541  542  543  544  target       state  week

In [38]:
X_test_encoded = pd.get_dummies(X_test, columns=['state'])
X_test_encoded = X_test_encoded.drop(columns=['target'])
y_test = X_test['target']
# Make predictions on the test data
y_pred = dt_classifier.predict(X_test_encoded)

# Evaluate the classifier
from sklearn.metrics import accuracy_score, classification_report

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Generate a classification report
report = classification_report(y_test, y_pred)
print("Classification Report:")
print(report)

Accuracy: 0.8875
Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.97      0.94     20998
           1       0.32      0.13      0.18      2282

    accuracy                           0.89     23280
   macro avg       0.61      0.55      0.56     23280
weighted avg       0.85      0.89      0.87     23280



Based on the provided classification report:

#### Accuracy: 
The overall accuracy of the model is 0.8875, indicating that approximately 88.75% of the predictions made by the model are correct.
#### Precision:
Precision for class 0 (negative class) is 0.91, which means that when the model predicts an instance as class 0, it is correct 91% of the time.
Precision for class 1 (positive class) is 0.32, indicating that the model's positive predictions for class 1 are correct only 32% of the time. This suggests that the model tends to make a significant number of false positive predictions for class 1.
#### Recall:
Recall for class 0 is 0.97, indicating that the model correctly identifies 97% of the actual instances of class 0.
Recall for class 1 is 0.13, meaning that the model captures only 13% of the actual instances of class 1. This implies that the model misses a large proportion of class 1 instances, leading to a low recall score.
#### F1-score:
The F1-score is the harmonic mean of precision and recall. For class 0, the F1-score is 0.94, indicating a good balance between precision and recall. However, for class 1, the F1-score is only 0.18, suggesting poor performance in correctly identifying instances of class 1.
#### Support:
The support column indicates the number of actual occurrences of each class in the test data. There are 20998 instances of class 0 and 2282 instances of class 1.
#### Conclusion:

The model performs well in identifying instances of class 0, as indicated by high precision, recall, and F1-score for class 0.
However, the model struggles significantly with class 1 predictions, with low precision, recall, and F1-score for class 1. This suggests that the model has difficulty distinguishing between class 0 and class 1 instances, and it tends to misclassify many instances of class 1 as class 0.

### Naive Bayes:

In [42]:
# ------------------------------- Optional:  Dictionaries Prints --------------------------------------------------


def print_likelihoods(likelihoods):
    """
    Prints the contents of the "likelihoods" dictionary in a user-friendly format.

    Args:
        likelihoods (dict): The "likelihoods" dictionary to be printed.
    """

    for feature, likelihood_dict in likelihoods.items():
        print(f"\nFeature: {feature}")
        for key, value in likelihood_dict.items():
            print(f"  - {key}: {value:.2f}")

# Example usage:
#likelihoods = {
    #"feature1": {"class1": 0.75, "class2": 0.25},
   # "feature2": {"class1": 0.60, "class2": 0.40}
#}

#print_likelihoods(likelihoods)


# _______________________________________________

def print_pred_priors(pred_priors):
    """
    Prints the contents of the "pred_priors" dictionary in a user-friendly format.

    Args:
        pred_priors (dict): The "pred_priors" dictionary to be printed.
    """

    for feature, priors_dict in pred_priors.items():
        print(f"\nFeature: {feature}")
        for value, probability in priors_dict.items():
            print(f"  - {value}: {probability:.2f}")




# _______________________________________________


def print_class_priors(class_priors):
    """
    Prints the contents of the "class_priors" dictionary in a user-friendly format.

    Args:
        class_priors (dict): The "class_priors" dictionary to be printed.
    """

    print("\nClass priors:")
    for class_label, probability in class_priors.items():
        print(f"  - {class_label}: {probability:.2f}")





# _______________________________________________




In [43]:
class NaiveBayes:

    def __init__(self):
        self.features = list  # List of feature names
        self.likelihoods = {}  # Likelihood of each feature value given a class
        self.class_priors = {}  # Prior probabilities of each class
        self.pred_priors = {}  # Prior probabilities of each feature value
        self.X_train = np.array  # Training features
        self.y_train = np.array  # Training labels
        self.train_size = int  # Number of training samples
        self.num_feats = int  # Number of features

    def fit(self, X, y):
        self.features = list(X.columns)
        self.X_train = X.values
        self.y_train = y.values
        self.train_size = len(y)
        self.num_feats = X.shape[1]

        for feature in self.features:
            self.likelihoods[feature] = {}
            self.pred_priors[feature] = {}
            for feat_val in np.unique(X[feature]):
                self.pred_priors[feature][feat_val] = 0
                for outcome in np.unique(y):
                    self.likelihoods[feature][f"{feat_val}_{outcome}"] = 0
                    self.class_priors[outcome] = 0

        self._calc_class_prior()
        self._calc_likelihoods()
        self._calc_predictor_prior()

        print("\n -------- Likelihoods Prints: ----------")
        print_likelihoods(self.likelihoods)
        print("\n -------- Pred_Priors Prints: ----------")
        print_pred_priors(self.pred_priors)
        print("\n -------- Class_Priors Prints: ---------")
        print_class_priors(self.class_priors)

    def _calc_class_prior(self):
        for outcome in np.unique(self.y_train):
            outcome_count = np.sum(self.y_train == outcome)
            self.class_priors[outcome] = outcome_count / self.train_size

    def _calc_likelihoods(self):
        for feature in self.features:
            for outcome in np.unique(self.y_train):
                outcome_rows = self.X_train[self.y_train == outcome]
                for feat_val in np.unique(self.X_train[:, self.features.index(feature)]):
                    feat_val_count = np.sum(outcome_rows[:, self.features.index(feature)] == feat_val)
                    self.likelihoods[feature][f"{feat_val}_{outcome}"] = feat_val_count / np.sum(
                        self.y_train == outcome)

    def _calc_predictor_prior(self):
        for feature in self.features:
            for feat_val in np.unique(self.X_train[:, self.features.index(feature)]):
                self.pred_priors[feature][feat_val] = np.sum(
                    self.X_train[:, self.features.index(feature)] == feat_val) / self.train_size

    def predict(self, X):
        results = []
        X = X.values
        for query in X:
            probs_outcome = {}
            for outcome in np.unique(self.y_train):
                prior = self.class_priors[outcome]
                likelihood = 1
                for feat, feat_val in zip(self.features, query):
                    likelihood *= self.likelihoods[feat][f"{feat_val}_{outcome}"]
                posterior = prior * likelihood
                probs_outcome[outcome] = posterior
            print(probs_outcome)
            result = max(probs_outcome, key=probs_outcome.get)
            results.append(result)
        return np.array(results)


In [44]:
from sklearn.model_selection import train_test_split

# Create a NaiveBayes cclassifier

nb_clf = NaiveBayes()

# Fit the classifier to the training data
nb_clf.fit(X, y)



 -------- Likelihoods Prints: ----------

Feature: 0
  - 0.0_0: 0.61
  - 0.0_1: 0.40
  - 1.0_0: 0.20
  - 1.0_1: 0.22
  - 2.0_0: 0.08
  - 2.0_1: 0.13
  - 3.0_0: 0.04
  - 3.0_1: 0.08
  - 4.0_0: 0.02
  - 4.0_1: 0.05
  - 5.0_0: 0.01
  - 5.0_1: 0.03
  - 6.0_0: 0.01
  - 6.0_1: 0.02
  - 7.0_0: 0.01
  - 7.0_1: 0.02
  - 8.0_0: 0.00
  - 8.0_1: 0.01
  - 9.0_0: 0.00
  - 9.0_1: 0.01
  - 10.0_0: 0.00
  - 10.0_1: 0.01
  - 11.0_0: 0.00
  - 11.0_1: 0.00
  - 12.0_0: 0.00
  - 12.0_1: 0.00
  - 13.0_0: 0.00
  - 13.0_1: 0.00
  - 14.0_0: 0.00
  - 14.0_1: 0.00
  - 15.0_0: 0.00
  - 15.0_1: 0.00
  - 16.0_0: 0.00
  - 16.0_1: 0.00
  - 17.0_0: 0.00
  - 17.0_1: 0.00
  - 18.0_0: 0.00
  - 18.0_1: 0.00
  - 19.0_0: 0.00
  - 19.0_1: 0.00
  - 20.0_0: 0.00
  - 20.0_1: 0.00
  - 21.0_0: 0.00
  - 21.0_1: 0.00
  - 22.0_0: 0.00
  - 22.0_1: 0.00
  - 23.0_0: 0.00
  - 23.0_1: 0.00
  - 24.0_0: 0.00
  - 24.0_1: 0.00
  - 25.0_0: 0.00
  - 25.0_1: 0.00
  - 26.0_0: 0.00
  - 26.0_1: 0.00
  - 27.0_0: 0.00
  - 27.0_1: 0.00
  - 28.0_0: 0.